File: 1. Fibonacci Numbers.cpp
// 1. Fibonacci Numbers
// Topic: Fibonacci Numbers
// Description: Compute the nth Fibonacci number using recursion with memoization to avoid
// exponential recomputation. Input: n (integer >= 0) Output: nth Fibonacci number

#include <bits/stdc++.h>
using namespace std;

// Memoization starts here: using vector<long long> memo
long long fib(int n, vector<long long>& memo)
{
    if (n <= 1)
        return n;
    if (memo[n] != -1)
        return memo[n];
    memo[n] = fib(n - 1, memo) + fib(n - 2, memo);
    return memo[n];
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    if (!(cin >> n))
        return 0;
    vector<long long> memo(n + 1, -1);
    cout << fib(n, memo) << '\n';
    return 0;
}

// Memoization process started at the fib function where memo vector is checked and written.
// Time Complexity: O(n)
// Space Complexity: O(n) (memoization array + recursion stack)

File: 1. Fibonacci Numbers.py
# 1. Fibonacci Numbers
# Topic: Fibonacci Numbers
# Description: Compute the nth Fibonacci number using recursion with memoization to avoid exponential recomputation.
# Input: n (integer >= 0)
# Output: nth Fibonacci number

import sys
from functools import lru_cache

# Memoization starts here: lru_cache on fib
@lru_cache(maxsize=None)
def fib(n: int) -> int:
    if n <= 1:
        return n
    return fib(n-1) + fib(n-2)

if __name__ == '__main__':
    data = sys.stdin.read().strip().split()
    if not data:
        sys.exit(0)
    n = int(data[0])
    print(fib(n))

# Memoization process started at fib decorated with lru_cache.
# Time Complexity: O(n)
# Space Complexity: O(n) (cache + recursion stack)
